source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1237 @@
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/mesh_deformation/fastscape.h>
+#include <aspect/geometry_model/box.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <aspect/postprocess/visualization.h>
+#include <ctime>
+
+namespace aspect
+{
+  namespace MeshDeformation
+  {
+
+    template <int dim>
+    void
+    FastScape<dim>::initialize ()
+    {
+      AssertThrow(Plugins::plugin_type_matches<const GeometryModel::Box<dim>>(this->get_geometry_model()),
+                  ExcMessage("Fastscape can only be run with a box model"));
+
+      const GeometryModel::Box<dim> *geometry
+        = dynamic_cast<const GeometryModel::Box<dim>*> (&this->get_geometry_model());
+
+      // Find the id associated with the top boundary and boundaries that call mesh deformation.
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      const std::set<types::boundary_id> boundary_ids
+        = this->get_mesh_deformation_handler().get_active_mesh_deformation_boundary_indicators();
+
+      // Get the deformation type names called for each boundary.
+      /*
+       * TODO: Why does this only work if I declare it in the header file?
+       * If I declare it here, then I get a no match operand[] error when trying to get the
+       * names variable.
+       */
+      mesh_deformation_boundary_indicators_map
+        = this->get_mesh_deformation_handler().get_active_mesh_deformation_names();
+
+      // Loop over each mesh deformation boundary, and make sure FastScape is only called on the surface.
+      for (std::set<types::boundary_id>::const_iterator p = boundary_ids.begin();
+           p != boundary_ids.end(); ++p)
+        {
+          const std::vector<std::string> names = mesh_deformation_boundary_indicators_map[*p];
+          for (unsigned int i = 0; i < names.size(); ++i )
+            {
+              if (names[i] == "fastscape")
+                AssertThrow((*p == relevant_boundary),
+                            ExcMessage("Fastscape can only be called on the surface boundary."));
+            }
+        }
+
+      // Initialize parameters for restarting fastscape
+      restart = this->get_parameters().resume_computation;
+      restart_step = 0;
+
+      // second is for maximum coordinates, first for minimum.
+      for (unsigned int i=0; i<dim; ++i)
```

Comment:
Note that when the origin is not zero, the maximum coordinates are origin+extents